### DISCUSSION TITLE
Development of v0.4.9.4

### DISCUSSION THREAD

**Discussion by git-developer** [C0rn3j/sc-controller#36] [General]
The build of v0.4.9.3 [failed](https://github.com/C0rn3j/sc-controller/actions/runs/11086725236). This might be related to the python version: it fails on bullseye, focal, jammy (python <3.11) and works on bookworm, trixie, noble (python >= 3.11). If you'd like to debug locally, you can run:
```sh
$ docker build -o build-output --build-arg BASE_CODENAME=focal .
```

The first failing commit is [b3edd80](https://github.com/C0rn3j/sc-controller/commit/b3edd804552e72d8960439a7942e183fc5a24768).

**Comment by C0rn3j** [comment]
focal doesn't have the build deps in the repo, jammy fails on the following error and noble just works with this.

Got it a bit further, but I don't have time at the moment to figure out why jammy refuses to use the wheel files atm.  

Maybe we can get away by installing `installer` from `pip` if you want to try that.

```bash
docker build -o build-output --build-arg BASE_CODENAME=jammy .
```

```dockerfile
ARG BASE_OS=ubuntu
ARG BASE_CODENAME=noble
FROM $BASE_OS:$BASE_CODENAME AS build-stage

# Download build dependencies
RUN <<EOR
  set -eu

  apt-get update
  apt-get install -y --no-install-recommends \
      gcc \
      librsvg2-bin \
      linux-headers-generic \
      python3-dev \
      python3-setuptools \
      python-is-python3 \
      python3-build \
      python3-venv \
      python3-poetry \
      python3-installer
  apt-get clean && rm -rf /var/lib/apt/lists/*
EOR
# Prepare working directory and target
COPY . /work
WORKDIR /work
ARG TARGET=/build/usr

# Build and install
RUN <<EOR
  set -eu
  #pip install poetry
  echo $(command -v python3)
  python -m build --wheel
  python -m installer --destdir="${TARGET}" dist/*.whl

  # Provide input-event-codes.h as fallback for runtime systems without linux headers
  cp -a \
      "$(find /usr -type f -name input-event-codes.h -print -quit)" \
      "$(find "${TARGET}" -type f -name uinput.py -printf '%h\n' -quit)"

  # Create short name symlinks for static libraries
  suffix=".cpython-*-$(uname -m)-linux-gnu.so"
  find "${TARGET}" -type f -path "*/site-packages/*${suffix}" \
    | while read -r path; do ln -sfr "${path}" "${path%${suffix}}.so"; done

  # Put AppStream metadata to required location according to https://wiki.debian.org/AppStream/Guidelines
  metainfo=/build/usr/share/metainfo
  mkdir -p "${metainfo}"
  cp -a scripts/sc-controller.appdata.xml "${metainfo}"

  # Convert icon to png format (required for icons in .desktop file)
  iconpath="${TARGET}/share/icons/hicolor/512x512/apps"
  mkdir -p "${iconpath}"
  rsvg-convert --background-color none -o "${iconpath}/sc-controller.png" images/sc-controller.svg
EOR

# Store build metadata
ARG TARGETOS TARGETARCH TARGETVARIANT
RUN export "TARGETMACHINE=$(uname -m)" && printenv | grep ^TARGET >>/build/.build-metadata.env

# Keep only files required for runtime
FROM scratch AS export-stage
COPY --from=build-stage /build /
```

```bash
2.887 removing build/bdist.linux-x86_64/wheel
2.943 * Creating virtualenv isolated environment...
2.943 * Installing packages in isolated environment... (setuptools>=60, wheel)
2.943 * Getting dependencies for wheel...
2.943 * Building wheel...
2.943 Successfully built sccontroller-0.4.9.3-cp310-cp310-linux_x86_64.whl
2.985 usage: python -m installer [-h] [--destdir path] [--compile-bytecode level]
2.985                            [--no-compile-bytecode]
2.985                            wheel
2.985 python -m installer: error: unrecognized arguments: dist/sccontroller-0.4.9.3-cp312-cp312-linux_x86_64.whl
```

**Reply by git-developer** [reply]
A build of your `Dockerfile` does not cancel on my machine (didn't test the application or build an AppImage, though).

Notice the different filenames in your output:
> Successfully built sccontroller-0.4.9.3-**cp310-cp310**-linux_x86_64.whl

vs.
> error: unrecognized arguments: dist/sccontroller-0.4.9.3-**cp312-cp312**-linux_x86_64.whl

**Reply by C0rn3j** [reply]
> A build of your Dockerfile does not cancel on my machine 

You're building for `jammy` as per the command I posted?

I do not understand why YOUR system would be any different from MY system, it is docker.

I do have 3.12 on host, but I have no clue how the installer module would know about that.  
pyproject has py38 as a target too.

Actually, the Ubuntu image should have `3.10.12`...which still does not explain why it works for you, if one of us has some cached docker image.

Can you repro after `docker system prune -a` ? Careful, it removes quite a bit.

EDIT: I think it is copying my prebuilt wheel files from my devel dir.
EDIT2: Fixed that.

**Reply by C0rn3j** [reply]
Fixed Docker builds in https://github.com/C0rn3j/sc-controller/commit/b312dc615c435bfe5596b3c83744a46be9bd0adc + https://github.com/C0rn3j/sc-controller/commit/97a6de4f9102bdea3463aa116faff3070211b7e0

I have partially reverted your indent changes from https://github.com/C0rn3j/sc-controller/commit/cf39960dce139500fe0cabf5370d9b3b7fef3f0e#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557.  
Kept the level, but switched back to tabs, as different people prefer different tab-widths, and we can't accommodate that with spaces. 

This remains broken on Focal which will need to start pulling deps from pip I suppose, as they aren't packaged in the distro even in ancient versions.

The appimages too will have to start pulling `python3-build`, `python3-installer` and `python3-poetry` and stop pointing to setup.py and use build and installer modules instead.

I'll switch the AppImages eventually when I have a bit more time, but I suspect you might beat me to it.

**Reply by git-developer** [reply]
I'm on it.

**Reply by git-developer** [reply]
The build output differs significantly. The main executable `usr/bin/scc` (`scripts/scc` in the sources) is missing. I found it's commented out in the TOML file. Why? When I include it, the application fails to start. Can you help?

**Reply by git-developer** [reply]
```
/tmp/squashfs-root/runtime/compat$ cat ../../usr/bin/scc
#! usr/bin/python
# -*- coding: utf-8 -*-
import re
import sys
from scc.bin.scc import main
if __name__ == "__main__":
    sys.argv[0] = re.sub(r"(-script\.pyw|\.exe)?$", "", sys.argv[0])
    sys.exit(main())

/tmp/squashfs-root/runtime/compat$ ../../AppRun 
Traceback (most recent call last):
  File "/tmp/squashfs-root/usr/bin/scc", line 5, in <module>
    from scc.bin.scc import main
ModuleNotFoundError: No module named 'scc.bin.scc'
```

**Reply by git-developer** [reply]
Adding `scc = "scc.scripts:main"` to the TOML fixes it for focal, but noble fails.

**Reply by git-developer** [reply]
Noble has everything under `usr/local/`, but the application is expecting files in `usr/` instead. This is a breaking change. How to fix? Maybe the setup.py args `--single-version-externally-managed --home "${TARGET}"` are involved.

**Reply by C0rn3j** [reply]
/usr/local is definitely wrong. Unsure why it gets put there.  
As a note, the docker image uses `/usr/usr` which is wrong but it still somehow ends up working.

My AUR package builds just fine, so that's weird.  
```
sc-controller-git /usr/bin/sc-controller
```

I have disabled `scc` as I've fixed the executables to be... executable.  
They are *supposed* to be present in the /usr/bin and perfectly usable.

`scc` was a hack to launch other executables/modules through it - "avoiding bin pollution".  
It should not be needed anymore, but it looks like the AppImages depend on it hard.

Looks like these need to be ideally changed, or we can include the hack script for now in the AppImages, if it still works fine with it:

```bash
[0] % grep -R /usr/bin/scc
AppImageBuilder.ubuntu-focal.yml:    exec_args: -c "import os, sys; os.execvp('$APPDIR/usr/bin/scc', ['$APPDIR/usr/bin/scc'] + (sys.argv[1:] if len(sys.argv) > 1 else ['gui']))" $@
AppImageBuilder.ubuntu-focal.yml:    sed -i "s:Exec=.*:Exec=./usr/bin/scc gui:g" "${desktop}"
AppImageBuilder.yml:    exec_args: -c "import os, sys; os.execvp('$APPDIR/usr/bin/scc', ['$APPDIR/usr/bin/scc'] + (sys.argv[1:] if len(sys.argv) > 1 else ['gui']))" $@
AppImageBuilder.yml:    sed -i "s:Exec=.*:Exec=./usr/bin/scc gui:g" "${desktop}"
scripts/appimage-AppRun.sh:python3 ${APPDIR}/usr/bin/scc dependency-check &>/tmp/scc.depcheck.$$.txt \
scripts/appimage-AppRun.sh:python3 ${APPDIR}/usr/bin/scc $ARG1 $@
AppImageBuilder.debian.yml:    exec_args: -c "import os, sys; os.execvp('$APPDIR/usr/bin/scc', ['$APPDIR/usr/bin/scc'] + (sys.argv[1:] if len(sys.argv) > 1 else ['gui']))" $@
AppImageBuilder.debian.yml:    sed -i "s:Exec=.*:Exec=./usr/bin/scc gui:g" "${desktop}"
AppImageBuilder.debian-bullseye.yml:    exec_args: -c "import os, sys; os.execvp('$APPDIR/usr/bin/scc', ['$APPDIR/usr/bin/scc'] + (sys.argv[1:] if len(sys.argv) > 1 else ['gui']))" $@
AppImageBuilder.debian-bullseye.yml:    sed -i "s:Exec=.*:Exec=./usr/bin/scc gui:g" "${desktop}"
```

**Reply by git-developer** [reply]
AppImages can be changed, as long as `daemon debug`, `gui` and `dependency-check` can be started somehow. But we should fix the `local` problem first.

**Reply by git-developer** [reply]
The python module `installer` behaves differently:
* Focal: installs to `usr`
* Jammy & Noble: installs to `usr/local`. Supports an arg `--prefix` to change `usr`, but `local` is always suffixed.

**Reply by C0rn3j** [reply]
I suppose the problem might lie in the variable that is passed to `--destdir`, is it the same between the distros?

**Reply by git-developer** [reply]
Yes, it is `/build` in all cases.

**Reply by git-developer** [reply]
My updates are in branch [dev](https://github.com/git-developer/sc-controller/tree/dev).

**Reply by git-developer** [reply]
`installer` is using the paths from `sysconfig`, and these are baked in by the distro. Didn't find a way to change that. But using `pip` in a dedicated venv for installation makes a difference. `local` directory is gone. But it's still broken: all python shebangs are pointing to the venv directory.

**Reply by git-developer** [reply]
Added a patch for the shebangs. And a workaround for a [bug](https://bugs.launchpad.net/ubuntu/+source/python-build/+bug/1992108) in Jammy. Now the build and tests succeed, and all AppImages are able to open the gui.

**Reply by C0rn3j** [reply]
> And a workaround for a [bug](https://bugs.launchpad.net/ubuntu/+source/python-build/+bug/1992108) in Jammy.

Looks like correct thing there would be to enable `jammy-proposed` repo in apt sources, update python-build and tell upstream in the bug report whether it fixes the issue or not, as OP there ignored them for half a year.

If you want to spend the time on that of course.

**Reply by git-developer** [reply]
Concerning the `scc` binary: I wouldn't remove it for 2 reasons. First, it is an incompatible change that was not announced. Users who call it today will fail. Second, it is a well-known pattern to have a single tool as facade for different functionalities. It's easier for a user to have one single tool than to remember a bunch of them.

**Reply by git-developer** [reply]
> ... enable `jammy-proposed` repo in apt sources ...

I will think about that. I agree that using the proposed repo is cleaner. Currently, the `Dockerfile` can be used for all distros. The proposed repo requires an explicit check for jammy, and the URL differs per arch 🙄

**Reply by git-developer** [reply]
BTW, I skipped the `poetry` dependency, it's not used.

**Reply by git-developer** [reply]
And once more concerning the `scc` script: AppImages actually _require_ a single executable because that's what they run when called. So we can't avoid the facade.

**Reply by git-developer** [reply]
Do you know whether the shebang patching is OK or if there's a cleaner way, @C0rn3j?

**Reply by C0rn3j** [reply]
Does `scc` actually work for you outside of AppImages?

Because when I was testing it, it always failed to pass parameters, which was why I just fixed the binaries instead.

> > ... enable `jammy-proposed` repo in apt sources ...
> 
> I will think about that. I agree that using the proposed repo is cleaner. Currently, the `Dockerfile` can be used for all distros. The proposed repo requires an explicit check for jammy, and the URL differs per arch 🙄

The idea would be that upstream fixes their package and we won't have to carry any workarounds

> BTW, I skipped the `poetry` dependency, it's not used.

Pretty sure it failed on Jammy for me without it. I could be wrong about it being necessary, if you're sure, sure.

EDIT: ah, you removed the installer module.

> Do you know whether the shebang patching is OK or if there's a cleaner way, @C0rn3j?

Wouldn't retaining installer module and specifying a prefix help here?  
https://installer.pypa.io/en/stable/cli/installer/

Maybe we should switch to wheel instead?  
https://packaging.python.org/en/latest/key_projects/#wheel

Actually looks like what you're doing is same/similar?  
https://wheel.readthedocs.io/en/latest/user_guide.html#installing-wheels

Is there a reason to not source the activate file of venv and just running `pip` instead of `.env/bin/pip install`?

Maybe we should just upgrade pip from pip in and everything else in the venv/build env, and then upgrade setuptools and whatever else that's needed to latest to ensure our envs are less unique instead of having workarounds for each weird distro version.

**Reply by git-developer** [reply]
> Does scc actually work for you outside of AppImages?

I never tried, I always use an AppImage to avoid the dependency hell. But there should be no difference if `scc` is called inside or outside an AppImage. In the end, it's just a zipped directory (you can unpack it using arg `--appimage-extract`) with a `AppRun` binary doing some lib-loading-foo and delegating to the main executable.

> retaining installer module and specifying a prefix

I tried, it didn't help. As written [above](#discussioncomment-10832180), `local` was suffixed always. Please try yourself, maybe I've overseen something. I never got in touch with installer, wheels and the like. I appreciate everything that slims down the process.

> Is there a reason to not source the activate file

No, I just chose it because it's a single command. Would that improve / simplify something?

> Maybe we should just upgrade pip ...

Sounds great! I'm happy that the build is working now, so we can use the CI to ensure any refactoring does not break anything. Can you prepare something in a separate branch? Once it's ready to try, we can simply run a manual GitHub Action build and use the artifacts for local tests.

**Reply by C0rn3j** [reply]
>  I never got in touch with installer, wheels and the like. I appreciate everything that slims down the process.

That makes two people learning as they go :D 

> there should be no difference if scc is called inside or outside an AppImage.

The problem is that migrating to pyproject.toml and the associated build process made the executables be a generated script pointing to a function, as opposed to the executables just being executed as is.

That's why I had to mark them as executable, edit some of them to have a main function and why I was able to throw out scc (or so I thought I could throw it).

> No, I just chose it because it's a single command. Would that improve / simplify something?

One could accidentally run pip after it and install thing into the host by accident, but if it works as is, other than that, perhaps not?

> Sounds great! I'm happy that the build is working now, so we can use the CI to ensure any refactoring does not break anything.

Agree. Let's push that in a separate PR. I'll get to it some time in the future™, hopefully by next week or something.

It's going to be nice as even Arch has setuptools out of date by a few months at the moment.

**Comment by git-developer** [comment]
I tested v0.4.9.4, both daemon and GUI work great. I created issues for a few error messages that I found in the log. They didn't impact the user experience / program behavior.
